### ISSUE TITLE
:substitute hangs if inccommand is set to split or nosplit

### ISSUE THREAD

**Issue by e-kwsm** [neovim/neovim#12919]
- `nvim --version`: v0.5.0-nightly-103-gf34eeba2d
- `vim -u DEFAULTS` (version: ) behaves differently? n/a
- Operating system/version: Debian 10.5
- Terminal name/version: GNOME Terminal 3.30.2 using VTE 0.54.2 +GNUTLS
- `$TERM`: `xterm-256color`

### Steps to reproduce using `nvim -u NORC`

```
echo 1 | nvim -u NORC -c 'set icm=split' -
:s/1/2
```

Also happens for `set icm=nosplit`.

### Actual behaviour

- `:s/1/`—neovim works correctly, `1` disappears
- `:s/1/2`—`2` is not shown and neovim hangs

Bisect reveals that 54ce1010e86cbb29067aabfc332dab59b5a88edb is the first bad commit.

### Expected behaviour

`1` is replaced with `2`.

**Comment by clason** [comment]
ASAN/UBSAN backtrace (from Gitter yesterday, @bfredl):
```
=================================================================
==78050==ERROR: AddressSanitizer: heap-use-after-free on address 0x61a000017a88 at pc 0x00010d1721c8 bp 0x7ffee3f794f0 sp 0x7ffee3f794e8
WRITE of size 8 at 0x61a000017a88 thread T0
    #0 0x10d1721c7 in u_savecommon undo.c:439
    #1 0x10d176f50 in u_savesub undo.c:246
    #2 0x10c2f5fa1 in do_sub ex_cmds.c:4037
    #3 0x10c2e7363 in ex_substitute ex_cmds.c:5816
    #4 0x10c36a595 in do_one_cmd ex_docmd.c:1949
    #5 0x10c3531b1 in do_cmdline ex_docmd.c:589
    #6 0x10c47078d in command_line_changed ex_getln.c:2228
    #7 0x10c484a3f in command_line_handle_key ex_getln.c:2150
    #8 0x10c46b2ee in command_line_execute ex_getln.c:1442
    #9 0x10cfcb32d in state_enter state.c:69
    #10 0x10c422651 in command_line_enter ex_getln.c:810
    #11 0x10c41db9f in getcmdline ex_getln.c:2305
    #12 0x10c424ee7 in getexline ex_getln.c:2484
    #13 0x10c3524f2 in do_cmdline ex_docmd.c:507
    #14 0x10c984e99 in nv_colon normal.c:4625
    #15 0x10c96e759 in normal_execute normal.c:1138
    #16 0x10cfcb32d in state_enter state.c:69
    #17 0x10c929375 in normal_enter normal.c:463
    #18 0x10c6c2748 in main main.c:557
    #19 0x7fff20d18950 in start+0x0 (libdyld.dylib:x86_64+0x15950)

0x61a000017a88 is located 8 bytes inside of 1216-byte region [0x61a000017a80,0x61a000017f40)
freed by thread T0 here:
    #0 0x10e31f326 in wrap_free+0xa6 (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x49326)
    #1 0x10c8409b4 in xfree memory.c:119
    #2 0x10d1a84a0 in u_freeentries undo.c:2822
    #3 0x10d1791d7 in u_freeheader undo.c:2755
    #4 0x10d1778bb in u_freebranch undo.c:2774
    #5 0x10d18bdf6 in u_undo_and_forget undo.c:1724
    #6 0x10c2e76ec in ex_substitute ex_cmds.c:5826
    #7 0x10c36a595 in do_one_cmd ex_docmd.c:1949
    #8 0x10c3531b1 in do_cmdline ex_docmd.c:589
    #9 0x10c47078d in command_line_changed ex_getln.c:2228
    #10 0x10c484a3f in command_line_handle_key ex_getln.c:2150
    #11 0x10c46b2ee in command_line_execute ex_getln.c:1442
    #12 0x10cfcb32d in state_enter state.c:69
    #13 0x10c422651 in command_line_enter ex_getln.c:810
    #14 0x10c41db9f in getcmdline ex_getln.c:2305
    #15 0x10c424ee7 in getexline ex_getln.c:2484
    #16 0x10c3524f2 in do_cmdline ex_docmd.c:507
    #17 0x10c984e99 in nv_colon normal.c:4625
    #18 0x10c96e759 in normal_execute normal.c:1138
    #19 0x10cfcb32d in state_enter state.c:69
    #20 0x10c929375 in normal_enter normal.c:463
    #21 0x10c6c2748 in main main.c:557
    #22 0x7fff20d18950 in start+0x0 (libdyld.dylib:x86_64+0x15950)

previously allocated by thread T0 here:
    #0 0x10e31f1dd in wrap_malloc+0x9d (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x491dd)
    #1 0x10c8406a2 in try_malloc memory.c:71
    #2 0x10c84087c in xmalloc memory.c:105
    #3 0x10d16fd8a in u_savecommon undo.c:377
    #4 0x10d1a4f7e in u_force_get_undo_header undo.c:3031
    #5 0x10c4b8e58 in u_extmark_set extmark.c:417
    #6 0x10c4b72ca in extmark_set extmark.c:129
    #7 0x10c4cb6bb in bufhl_add_hl_pos_offset extmark.c:774
    #8 0x10c30567e in show_sub ex_cmds.c:5747
    #9 0x10c2f9ced in do_sub ex_cmds.c:4224
    #10 0x10c2e7363 in ex_substitute ex_cmds.c:5816
    #11 0x10c36a595 in do_one_cmd ex_docmd.c:1949
    #12 0x10c3531b1 in do_cmdline ex_docmd.c:589
    #13 0x10c47078d in command_line_changed ex_getln.c:2228
    #14 0x10c484a3f in command_line_handle_key ex_getln.c:2150
    #15 0x10c46b2ee in command_line_execute ex_getln.c:1442
    #16 0x10cfcb32d in state_enter state.c:69
    #17 0x10c422651 in command_line_enter ex_getln.c:810
    #18 0x10c41db9f in getcmdline ex_getln.c:2305
    #19 0x10c424ee7 in getexline ex_getln.c:2484
    #20 0x10c3524f2 in do_cmdline ex_docmd.c:507
    #21 0x10c984e99 in nv_colon normal.c:4625
    #22 0x10c96e759 in normal_execute normal.c:1138
    #23 0x10cfcb32d in state_enter state.c:69
    #24 0x10c929375 in normal_enter normal.c:463
    #25 0x10c6c2748 in main main.c:557
    #26 0x7fff20d18950 in start+0x0 (libdyld.dylib:x86_64+0x15950)

SUMMARY: AddressSanitizer: heap-use-after-free undo.c:439 in u_savecommon
Shadow bytes around the buggy address:
  0x1c3400002f00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f30: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x1c3400002f40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x1c3400002f50: fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f60: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f80: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002f90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c3400002fa0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==78050==ABORTING
``

**Comment by erw7** [comment]
This should have been fixed by #12968. If it hasn't been fixed, please reopen this.
